staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go

```diff
@@ -248,17 +267,27 @@ func (t *envelopeTransformer) doEncode(request *kmstypes.EncryptedObject) ([]byt
 	if err := validateEncryptedObject(request); err != nil {
 		return nil, err
 	}
-	return proto.Marshal(request)
+
+	buffer := t.encryptedObjectDecoder.bufferPool.Get().(*proto.Buffer)
+	defer t.encryptedObjectDecoder.bufferPool.Put(buffer)
+
+	if err := buffer.Marshal(request); err != nil {
+		return nil, err
+	}
+
+	return buffer.Bytes(), nil
 }
 
 // doDecode decodes the byte array to an EncryptedObject.
 func (t *envelopeTransformer) doDecode(originalData []byte) (*kmstypes.EncryptedObject, error) {
-	o := &kmstypes.EncryptedObject{}
-	if err := proto.Unmarshal(originalData, o); err != nil {
-		return nil, err
-	}
-	// validate the EncryptedObject
-	if err := validateEncryptedObject(o); err != nil {
+	buffer := t.encryptedObjectDecoder.bufferPool.Get().(*proto.Buffer)
+	defer t.encryptedObjectDecoder.bufferPool.Put(buffer)
+	buffer.SetBuf(originalData)
```

Comment:
I think we can remove the use of `bufferPool` in this path because the `SetBuf` basically makes the buffer not re-use any pre-allocated state (keep using `objectPool`).